### PR TITLE
[GRIFFIN-316] Fix job exception handling

### PR DESCRIPTION
**Pull request by PnPie**
**What changes were proposed in this pull request?**

Currently we are using Try instance to represent the results of a DQ job, whether succeeded or failed. But as we are only wrapping the Boolean result by applying "Try" at the most outside level, the underlying failure would not be able to caught and it would always return "Success" even if exception got.

This is to modify all the underlying execute/doExecute methods of a DQ job, by handling exception with "Try" instances so that it could be passed properly to users when things get wrong.

**Does this PR introduce any user-facing change?**
No.

**How was this patch tested?**
Griffin test suite.

### CODE REVIEW

**Comment by wankunde** [comment]
Hi, @PnPie 

Could you format your code according to  `https://github.com/apache/griffin/blob/master/griffin-doc/dev/code-style.md`？

**Comment by PnPie** [comment]
Hi @wankunde 
I pushed some changes for reformatting codes.

File: measure/src/main/scala/org/apache/griffin/measure/job/DQJob.scala [hunk 1]
```diff
@@ -25,8 +27,16 @@ case class DQJob(dqSteps: Seq[DQStep]) extends Serializable {
   /**
    * @return execution success
    */
-  def execute(context: DQContext): Boolean = {
-    dqSteps.forall(dqStep => dqStep.execute(context))
+  def execute(context: DQContext): Try[Boolean] = {
+    val tmp = dqSteps.map(dqStep => dqStep.execute(context))
```
Comment (chitralverma) [inline measure/src/main/scala/org/apache/griffin/measure/job/DQJob.scala]: mapping the elements of `Seq[_]` is not lazy. If there are issues while `.execute`, they will not be handled.

Also, a different variable name can be used in place of `tmp`.

**Comment by wankunde** [comment]
Hi, @PnPie @chitralverma 

Now some `TransformStep` may execute incorrectly ,but return true.
In my opinion, we should catch those exceptions and return false in `doExecute(context: DQContext)` method.

File: measure/src/main/scala/org/apache/griffin/measure/job/DQJob.scala (hunk 1 above)
Comment (PnPie) [reply measure/src/main/scala/org/apache/griffin/measure/job/DQJob.scala]: I changed the `.execute` method inside and they return a `Try[Boolean]` instance, which wraps the failure potentially be raised.
In the test case added, the spark sql `AnalysisException` is actually raised here from `.execute` method.

File: measure/src/main/scala/org/apache/griffin/measure/step/transform/TransformStep.scala [hunk 2]
```diff
@@ -67,15 +66,8 @@ trait TransformStep extends DQStep {
...
-    if (prepared) {
-      val res = doExecute(context)
-      info(threadName + " end transform step : \n" + debugString())
-      res
-    } else {
-      error("Parent transform step failed!")
-      false
-    }
+    val prepared = parentSteps.foldLeft(true)((ret, step) => ret && step.status == COMPLETE)
+    doExecute(context)
```
[the remark this answers is not in the document]
Comment (PnPie) [reply measure/src/main/scala/org/apache/griffin/measure/step/transform/TransformStep.scala]: Sorry I didn't finish it properly and just put the updates.

**Comment by guoyuepeng** [comment]
any updates?

**Comment by PnPie** [comment]
Hi all,
Do you think it's better to catch the exception in `TransformStep` and return false ? if so I can modify it accordingly.

**Comment by wankunde** [comment]
Hi, @PnPie 

I'm so sorry for  for not responding this PR in time.

Could you update this PR for for master branch.

**Comment by PnPie** [comment]
Hi @wankunde 
Yes I just merged master in.

**Comment by guoyuepeng** [comment]
> Hi @wankunde
> Yes I just merged master in.

sure, we will merge it after the CI successfully completed.

**Comment by PnPie** [comment]
Great ! Thanks @wankunde @guoyuepeng

**Comment by guoyuepeng** [comment]
@PnPie could you send me your personal email?
my email is guoyp@apache.org.

Thanks.

**Comment by PnPie** [comment]
@guoyuepeng Yes I sent you a message just now.

**Comment by guoyuepeng** [comment]
@PnPie 

2020-03-11 21:59:35.138 ERROR [org.apache.griffin.measure.job.BatchDQAppTest] - assertion failed: Connector is undefined or invalid
java.lang.AssertionError: assertion failed: Connector is undefined or invalid
	at scala.Predef$.assert(Predef.scala:170)
	at org.apache.griffin.measure.configuration.dqdefinition.DataSourceParam.validate(DQConfig.scala:100)
	at org.apache.griffin.measure.configuration.dqdefinition.DQConfig$$anonfun$validate$5.apply(DQConfig.scala:74)
	at org.apache.griffin.measure.configuration.dqdefinition.DQConfig$$anonfun$validate$5.apply(DQConfig.scala:74)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.apache.griffin.measure.configuration.dqdefinition.DQConfig.validate(DQConfig.scala:74)
	at org.apache.griffin.measure.configuration.dqdefinition.reader.ParamReader$class.validate(ParamReader.scala:43)
	at org.apache.griffin.measure.configuration.dqdefinition.reader.ParamFileReader.validate(ParamFileReader.scala:33)
	at org.apache.griffin.measure.configuration.dqdefinition.reader.ParamFileReader$$anonfun$readConfig$1.apply(ParamFileReader.scala:40)
	at org.apache.griffin.measure.configuration.dqdefinition.reader.ParamFileReader$$anonfun$readConfig$1.apply(ParamFileReader.scala:36)
	at scala.util.Try$.apply(Try.scala:192)
	at org.apache.griffin.measure.configuration.dqdefinition.reader.ParamFileReader.readConfig(ParamFileReader.scala:36)
	at org.apache.griffin.measure.Application$.readParamFile(Application.scala:127)
	at org.apache.griffin.measure.job.DQAppTest.initApp(DQAppTest.scala:55)
	at org.apache.griffin.measure.job.BatchDQAppTest$$anonfun$7.apply$mcV$sp(BatchDQAppTest.scala:155)
	at org.apache.griffin.measure.job.BatchDQAppTest$$anonfun$7.apply(BatchDQAppTest.scala:154)
	at org.apache.griffin.measure.job.BatchDQAppTest$$anonfun$7.apply(BatchDQAppTest.scala:154)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FlatSpecLike$$anon$1.apply(FlatSpecLike.scala:1682)
	at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
	at org.scalatest.FlatSpec.withFixture(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$class.invokeWithFixture$1(FlatSpecLike.scala:1679)
	at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
	at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
	at org.scalatest.FlatSpecLike$class.runTest(FlatSpecLike.scala:1692)
	at org.scalatest.FlatSpec.runTest(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
	at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:373)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:410)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
	at org.scalatest.FlatSpecLike$class.runTests(FlatSpecLike.scala:1750)
	at org.scalatest.FlatSpec.runTests(FlatSpec.scala:1685)
	at org.scalatest.Suite$class.run(Suite.scala:1147)
	at org.scalatest.FlatSpec.org$scalatest$FlatSpecLike$$super$run(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
	at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
	at org.scalatest.FlatSpecLike$class.run(FlatSpecLike.scala:1795)
	at org.apache.griffin.measure.job.DQAppTest.org$scalatest$BeforeAndAfterAll$$super$run(DQAppTest.scala:33)
	at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
	at org.apache.griffin.measure.job.DQAppTest.run(DQAppTest.scala:33)
	at org.scalatest.Suite$class.callExecuteOnSuite$1(Suite.scala:1210)
	at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1257)
	at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1255)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at org.scalatest.Suite$class.runNestedSuites(Suite.scala:1255)
	at org.scalatest.tools.DiscoverySuite.runNestedSuites(DiscoverySuite.scala:30)
	at org.scalatest.Suite$class.run(Suite.scala:1144)
	at org.scalatest.tools.DiscoverySuite.run(DiscoverySuite.scala:30)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1340)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1334)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1334)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1011)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1010)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1500)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1010)
	at org.scalatest.tools.Runner$.main(Runner.scala:827)
	at org.scalatest.tools.Runner.main(Runner.scala)

**Comment by PnPie** [comment]
I see that it failed because we have refactored Data Source configuration in latest commit #568, in which we renamed `connectors` to `connector`, and I just modified it.

**Comment by guoyuepeng** [comment]
ok, my local build passed.

**Comment by guoyuepeng** [comment]
CI passed.

**Pull request closed without merging**